### ISSUE TITLE
Build graphs imperatively

### ISSUE THREAD

**Issue by rolyp** [explorable-viz/fluid#1008]
Accumulating a graph “descriptor” in a monoid and then imperatively converting it into a graph may account for some of the current performance overhead. It may be worth just using the `ST` monad to build the graph directly, forgoing the intermediate data structure.

- [ ] New `bwdSlice` impl
- [ ] New `fwdSlice` impl (although: I don't think we use this any more!)
- [ ] New `graphEval` impl
- [ ] `Unit` rather than `Bool` in `VertexSet`
- [ ] Check performance with `Agriculture` field of `moving-average` example
- [ ] Experiment using `Set` typeclass (backed by `Dict`/foreign object) rather than `Data.Set` in `Graph` interface
  - [ ] Initial experiment with `sources` and `sinks`
  - [ ] If that is reasonable, then same for `vertices`
  - [ ] Revisit `sinks'` implementation?
- [ ] Unify two `vertices` implementations (`Set DVertex` in `Vertices` and `Set Vertex` in `GraphImpl`)

**Comment by rolyp** [comment]
As per suggestion from @jacobpake if we used `ST` to implement `bwdSlice` we could switch to a JS object to implement the visited set, which would also probably be a big performance improvement.

**Comment by jacobpake** [comment]
Other points of interest here would be:

`fromEdgeList` and its dependencies, specifically `sinks'`, `inMap`, `keys`, `outMap`.

`vertices` - need to track down which implementations are most impactful, but this appears often in the call tree.

Generally the implementations here and in `bwdSlice` look good to me, but maybe let down by the underlying data structures JS overhead

**Comment by rolyp** [comment]
@jacobpake Thanks. @JosephBond pointed out earlier today that for `vertices` (on values) we use the "map-singleton-or-const-mempty and then folding union" pattern, and that this might be improved by using the `Foldable` instance to build a list first and then turning that into a set in one go. Let's have a chat next week and think about what the low-hanging fruit are here! (I guess as we pick off the obvious wins, other bottlenecks will start to become more prominent.)

**Comment by rolyp** [comment]
There seem to be 2 obvious implementations of `Set` based on an underlying JS object. (1) an interface that looks pure and under the hood is indeed pure, e.g. `insert` returns a new object with an additional key (as per your current implementation I think), rather than doing an in-place edit. (2) similar but does an in-place edit.

(2) is clearly unsafe so we can't really do that. (1) is safe but from experience the asymptotics are bad. Insertion is complexity something like _O(n log n)_ because of the need to keep cloning the object. Even though copying an object is a super-fast operation, it does eventually become slow. We had very degenerate behaviour using this approach to build large graphs.

The remaining options are:
(3) `Data.Set`, staying within PureScript -- better asymptotics (_O(log n)_) but overall slow.
(4) Imperative set update in the `ST` monad -- safe and much faster.

I think (4) is therefore probably the road to go down, at least for graph building. And I'm also wondering whether we even need the visited set if we build the graph imperatively. I may be misunderstanding, but I think the visited set may only be needed because otherwise there isn't a way to see what nodes have been added to the graph already with the `WithGraphT` monad (which accumulates a list of edges which would be expensive to query). If we're building the graph as an object in `ST`, we can probably just treat the keys of the map as the visited set.

**Comment by jacobpake** [comment]
I tried in a few areas implementing Set as a Javascript [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

I think this only make senses where there is no deep copying needed i.e. for Set String / Set Int and the advantage comes from the native Set implementation (which is pretty fast these days) and avoiding some of the Purescript overhead (function calls, comparisons, object create/destroy) in hot loops.

The areas where this approach is actually useful is quite limited - mostly in the `bwdSlice` visited set - which gets us about 10% speedup vs current implementation - but less as the graph gets bigger unless we do unsafe mutation.

I think you are right that we can use the graph object keys to track what has been visited.

**Comment by rolyp** [comment]
Yes my guess is that the speedup for smallish examples will in larger examples be dominated by the slowdown associated with insertion having super-linear (rather than the usual sub-linear) time complexity. So let's pursue building the graph objects in `ST`, and see if we can also avoid building a separate `visited` set, which then might remove a lot of overhead whilst preserving the current asymptotic behaviour.

**Comment by jacobpake** [comment]
Replacing the Visited set with immutable Foreign Object keys was a 10x slowdown on graph building

**Comment by jacobpake** [comment]
Mutable Object in ST worked well

**Comment by rolyp** [comment]
Hi @jacobpake. As mentioned I'm off next week and you're off Tuesday. I added some notes from our conversation yesterday to the issue -- perhaps the goal for next week should be to wrap up this current reimplementation to whatever degree is achievable by end of the week. (Progress seems good so far!) I also added a todo about `graphEval` using the new infrastructure.

**Comment by rolyp** [comment]
Parked for now
